go/oauth2util/util.go

```diff
@@ -0,0 +1,75 @@
+package oauth2util
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+)
+
+// Default 3LO authorization handler. Prints the authorization URL on stdout
+// and reads the verification code from stdin.
+func defaultAuthorizeFlowHandler(authorizeUrl string) (string, error) {
+	// Print the url on console, let user authorize and paste the token back.
+	fmt.Printf("Go to the following link in your browser:\n\n   %s\n\n", authorizeUrl)
+	fmt.Println("Enter verification code: ")
+	var code string
+	fmt.Scanln(&code)
+	return code, nil
+}
+
+// NewClient create a new OAuth2 Client.
+//
+// key: a JSON string that represents either an OAuth client ID or a
+// service account key.
+// scope: A space separated scope codes per
+//     [OAuth 2.0 spec](https://tools.ietf.org/html/rfc6749).
+func NewClient(key []byte, scope []string, authorizeHandler func(string) (string, error)) (oauth2.TokenSource, error) {
+	var secret map[string]interface{}
+	if err := json.Unmarshal(key, &secret); err != nil {
+		return nil, err
+	}
+
+	if authorizeHandler == nil {
+		authorizeHandler = defaultAuthorizeFlowHandler
+	}
+
+	// TODO: support "web" client secret by using a local web server.
+	// According to the content in the json, decide whether to run three-legged
+	// flow (for client secret) or two-legged flow (for service account).
+	if _, ok := secret["installed"]; ok {
+		// When the secret contains "installed" field, it is a client secret. We
+		// will run a three-legged flow
+		conf, err := google.ConfigFromJSON(key, scope...)
+		if err != nil {
+			return nil, err
+		}
+
+		// In the authorize flow, user will paste a verification code back to console.
+		authUrl := conf.AuthCodeURL("", oauth2.AccessTypeOffline, oauth2.ApprovalForce)
+		code, err := authorizeHandler(authUrl)
+		if err != nil {
+			return nil, err
+		}
+
+		// The verify flow takes in the verification code from authorize flow, sends a
+		// POST request containing the code to fetch oauth token.
+		token, err := conf.Exchange(nil, code)
+		if err != nil {
+			return nil, err
+		}
+
+		return conf.TokenSource(nil, token), nil
+
+	} else if tokenType, ok := secret["type"]; ok && "service_account" == tokenType {
+		// If the token type is "service_account", we will run the two-legged flow
+		jwtCfg, err := google.JWTConfigFromJSON(key, scope...)
```

Comment:
I guess the client id json and the service account key json are defined by Google?